Flink 7077

EronWright opened the pull request:
## Description

_This PR extends #4555 - please disregard commits from Aug 16, 2017._ 

[FLINK-7077] [mesos] Implement task release to support dynamic scaling
    
    - SlotManager: fix for idleness tracking (`markIdle` shouldn't reset `idleSince` on every call)
    - ResourceManager: change `stopWorker` method to use `ResourceID`
    - ResourceManager: schedule callbacks from `ResourceManagerActions` onto main thread
    - MesosResourceManager: implement `stopWorker`
    - MesosResourceManager: fix for message routing from child actors to RM

This change added tests and can be verified as follows:
- `MesosResourceManagerTest::testStopWorker`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**

tillrohrmann commented:
Thanks for you contribution @EronWright. `MesosResourceManagerTest` seems to fail on Travis.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -960,12 +965,26 @@ public void handleError(final Exception exception) {
+				public void run() {
+					for (Map.Entry<ResourceID, WorkerRegistration<WorkerType>> entry : taskExecutors.entrySet()) {
+						if (entry.getValue().getInstanceID().equals(instanceId)) {
+							stopWorker(entry.getKey());
```
In the future we should make these ids being composed of each other. Then we should easily obtain the `ResourceID` from the `InstanceID`

EronWright replied:
I was thinking along those same lines - the slot manager deals with `InstanceID` mostly, and its log lines are tough to correlate with lower-level resource manager information that is `ResourceID` based.   It would be a nice improvement to make `InstanceID` a composite key that included `ResourceID`.

Regardless, I think the `stopWorker` method should use `ResourceID`, because `InstanceID` isn't a concept exposed to the RM subclasses.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/TaskManagerRegistration.java:
```diff
@@ -68,7 +68,9 @@ public boolean isIdle() {
-		idleSince = System.currentTimeMillis();
+		if (!isIdle()) {
+			idleSince = System.currentTimeMillis();
+		}
```
good catch!

EronWright replied:
The reason that the SM idleness unit tests didn't catch this is, `markIdle` is called when a slot report is received, which in practice occurs continuously but is tough to simulate.  Just making clear that there are idle tests but I didn't see an obvious way to improve them.

EronWright commented:
@tillrohrmann fixed the test issue and other noted issues.  Once #4555 is in, this is good to go.

The test issue was a race in RM acquiring leadership vs RPC call.   Added some code to `TestLeaderElectionService` to wait for leader ack and updated the base RM and Mesos RM to use it.
